helper/awsutil/error.go

```diff
@@ -0,0 +1,31 @@
+package awsutil
+
+import (
+	awsRequest "github.com/aws/aws-sdk-go/aws/request"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/logical"
+)
+
+// CheckAWSError will examine an error and convert to a logical error if
+// approrpiate. If no appropriate error is found, return nil
+func CheckAWSError(err error) error {
```

Comment:
Suggestions for better names for this function as well as `AppendLogicalError` are most welcome

Reply:
A couple of comments on this file:

- Do you need have CheckAWSError as a separate, exported function? The functionality seems like it could just be part of AppendLogicalError.
- Is multierror the best response pattern for AppendLogicalError, or would `errwrap` be more appropriate? I view the logical error as adding context to the underlying AWS error (very much an errwrap concept), vs. multierror which is often used to accumulate various errors together
- Depending on what you think of ^^^, maybe a single WrapAWSError would work.

Reply:
>Do you need have CheckAWSError as a separate, exported function?

`CheckAWSError` is for transforming an upstream error into a different error of our choosing. In this case (and only case I'm aware of), it's "AWS gave us `503` (or other throttle error) so we change that to a `BadGateway`). In short "do we want to change this error to something else?" If `yes`, that something gets returned. If `no`, then return nothing. Another way of doing it would be to return `(error, bool)`, with `bool` indicating if a change was made. If `false`, then the `error` returned is the same as the function received. 

The need for a separate method is shown in https://github.com/hashicorp/vault/blob/7b00112a52fa9c8b2a66b3f256a18c5a40ee242a/builtin/logical/aws/secret_access_keys.go#L218 , where we're already returning a error response, and we're using the returned `error` here to indicate a category of error. That error is later checked in [response_util](https://github.com/hashicorp/vault/blob/27b72e80abf9b236469800c05ee33d152cc95a12/logical/response_util.go#L96-L111) to control how Vault responds.

`AppendLogicalError` came up because in in [auth/aws/backend](https://github.com/hashicorp/vault/blob/7b00112a52fa9c8b2a66b3f256a18c5a40ee242a/builtin/credential/aws/backend.go#L237) we weren't using `logical.ErrorResponse`, so the original error wasn't going to be returned if I just used `CheckAWSError`. So, I wanted to return both errors, the original and the `logical` one, if necessary. 

My first approach used `errwrap` however, either it's not suited or I did it wrong, because I wasn't able to get `errwrap.Contains` to find the `logical.ErrUpstreamRateLimited` when wrapping the errors. Using `multierror` did work, and is compatible with `errwrap` and honestly made more sense to me than `errwrap`.